src/lib/core/Optional.h

```diff
@@ -107,8 +74,21 @@ class Optional
     static Optional<T> Value(const T & value) { return Optional(value); }
 
 private:
-    T mValue;       ///< Value IFF optional contains a value
-    bool mHasValue; ///< True IFF optional contains a value
+    struct None
+    {
+        static constexpr const std::size_t VariantId = 1;
+        bool operator==(const None &) const { return true; }
+    };
+
+    struct Some
+    {
+        static constexpr const std::size_t VariantId = 2;
+        Some(const T & value) : mValue(value) {}
+        bool operator==(const Some & other) const { return mValue == other.mValue; }
+        T mValue;
+    };
+
+    Variant<None, Some> mVariant;
```

Comment:
Are there any size implications on this? previous  code had sizeof(value) + sizeof(bool) as size (although alignment  may make it bigger) and new code may have variantid instead.

Reply:
Yes, mHasValue is replaced by VariantId. Other part is object T, same as before.